src/hal/src/command/mod.rs

```diff
@@ -15,53 +15,72 @@ pub use self::raw::RawCommandBuffer;
 pub use self::renderpass::*;
 pub use self::transfer::*;
 
-/// Thread-safe finished command buffer for submission.
-pub struct Submit<B: Backend, C>(pub(crate) B::CommandBuffer, PhantomData<C>);
-unsafe impl<B: Backend, C> Send for Submit<B, C> {}
+/// Trait indicating how many times a Submit can be submitted.
+pub trait Shot { }
+/// Indicates a Submit that can only be submitted once.
+pub enum OneShot { }
+impl Shot for OneShot { }
+/// Indicates a Submit that can be submitted multiple times.
+pub enum MultiShot { }
+impl Shot for MultiShot { }
 
-impl<B: Backend, C> Submit<B, C> {
+/// A trait representing a command buffer that can be added to a `Submission`.
+pub unsafe trait Submittable<B: Backend, C> {
     ///
-    pub unsafe fn new(buffer: CommandBuffer<B, C>) -> Self {
-        Submit(buffer.raw.clone(), PhantomData)
-    }
+    unsafe fn into_buffer(self) -> B::CommandBuffer;
```

Comment:
I would prefer returning a reference to a commandbuffer to avoid costly clones and adjust the queue submission interface accordingly.

Reply:
In order to enforce `OneTime` command buffers, the queue submission interface must take all `Submittable`s by value. Returning a reference would require them to be kept alive somehow.

Reply:
I though about something like [this](https://play.rust-lang.org/?gist=f518fc0bbb9c19d4df7018368c01bb22&version=stable)
I don't remember the exact internals regarding queue submission, but this should be possible.

Reply:
Oh, I never even realized that was possible! Cool!

Reply:
@msiglreith Hm, unfortunately the issue here is that the new structs would have to be generic, which would lead to really annoying signatures like `CommandBuffer<'a, B, C, L, SubmitOnce<B, C, L>>`.